File: app/src/main/java/com/aliya/base/sample/util/LifecycleFpsMonitorCompat.java
package com.aliya.base.sample.util;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;

/**
 * LifecycleFpsMonitor
 *
 * @author a_liYa
 * @date 2020/11/10 22:51.
 */
public final class LifecycleFpsMonitorCompat {

    public static void addObserver(final Lifecycle lifecycle, final Lifecycle.State state,
                                   final FpsMonitor.FpsListener listener) {
        lifecycle.addObserver(new LifecycleObserver() {

            @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
            public void onCreate() {
                if (Lifecycle.State.CREATED.isAtLeast(state)) {
                    FpsMonitor.get().registerMonitor(listener);
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_START)
            public void onStart() {
                if (Lifecycle.State.STARTED == state) {
                    FpsMonitor.get().registerMonitor(listener);
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
            public void onResume() {
                if (Lifecycle.State.RESUMED == state) {
                    FpsMonitor.get().registerMonitor(listener);
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
            public void onPause() {
                if (Lifecycle.State.RESUMED == state) {
                    FpsMonitor.get().unregisterMonitor(listener);
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
            public void onStop() {
                if (Lifecycle.State.STARTED == state) {
                    FpsMonitor.get().unregisterMonitor(listener);
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
            public void onDestroy() {
                if (Lifecycle.State.CREATED.isAtLeast(state)) {
                    FpsMonitor.get().unregisterMonitor(listener);
                }
            }
        });
    }

}

File: monitor/src/main/java/com/aliya/monitor/fps/LifecycleFpsMonitorCompat.kt
package com.aliya.monitor.fps

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import com.aliya.monitor.fps.FpsMonitor.FpsListener

/**
 * LifecycleFpsMonitor
 *
 * @author a_liYa
 * @date 2020/11/10 22:51.
 */
object LifecycleFpsMonitorCompat {
    fun addObserver(lifecycle: Lifecycle, state: Lifecycle.State,
                    listener: FpsListener) {
        lifecycle.addObserver(object : LifecycleObserver {
            @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
            fun onCreate() {
                if (Lifecycle.State.CREATED.isAtLeast(state)) {
                    FpsMonitor.registerMonitor(listener)
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_START)
            fun onStart() {
                if (Lifecycle.State.STARTED == state) {
                    FpsMonitor.registerMonitor(listener)
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
            fun onResume() {
                if (Lifecycle.State.RESUMED == state) {
                    FpsMonitor.registerMonitor(listener)
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
            fun onPause() {
                if (Lifecycle.State.RESUMED == state) {
                    FpsMonitor.unregisterMonitor(listener)
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
            fun onStop() {
                if (Lifecycle.State.STARTED == state) {
                    FpsMonitor.unregisterMonitor(listener)
                }
            }

            @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
            fun onDestroy() {
                if (Lifecycle.State.CREATED.isAtLeast(state)) {
                    FpsMonitor.unregisterMonitor(listener)
                }
            }
        })
    }
}